scripts/man/xrepo.1
.TH "xrepo" "1"
.SH NAME
xrepo \- cross-platform build utility based on Lua


.SH SYNOPSIS
.B xrepo
.RI [ action "] [" options ]


.SH DESCRIPTION
.B xrepo
is a lightweight cross-platform build utility based on Lua. It uses
.I xrepo.lua
to maintain project builds. Compared with
.IR makefile / CMakeLists.txt ,
the configuration syntax is more concise and intuitive. It is very friendly to
novices and can quickly get started in a short time. Let users focus more on
actual project development.


.SH ACTIONS

.TP
.B clean
Clear all package caches and remove all not\-referenced packages.

.TP
.B env
Set environment and execute command, or print environment.

.TP
.B export
Export the given packages.

.TP
.B fetch
Fetch library information of the given installed packages.

.TP
.B import
Import the given packages.

.TP
.B info
Show information of the given packages.

.TP
.B install
Install the given packages.

.TP
.B remove
Remove the given packages.

.TP
.B scan
Scan the given or all installed packages.

.TP
.B search
Search the given packages.

.TP
.B add\-repo
Add the given remote repository url.

.TP
.B list\-repo
List all remote repositories.

.TP
.B rm\-repo
Remove the given remote repository.

.TP
.B update\-repo
Update all local repositories from remote.


.SH OPTIONS

.TP
.B \-q, \-\-quiet
Quiet operation.

.TP
.B \-y, \-\-yes
Input yes by default if need user confirm.

.TP
.B \-\-root
Allow one to run xrepo as root.

.TP
.B \-v, \-\-verbose
Print lots of verbose information for users.

.TP
.B \-D, \-\-diagnosis
Print lots of diagnosis information.

.TP
.B \-\-version
Print the version number and exit.

.TP
.B \-h, \-\-help
Print this help message and exit.


.SH AUTHOR
.B xrepo
is written by
.MT waruqi@\:gmail.com
ruki
.ME .

This manual page was written by
.MT mmyangfl@\:gmail.com
Yangfl
.ME
for the Debian Project (and may be used by others).


CHANGELOG.md
# Changelog ([中文](#中文))

## master (unreleased)

### New features

* [#4250](https://github.com/xmake-io/xmake/pull/4250): Improve link mechanism and order
* [#1438](https://github.com/xmake-io/xmake/issues/1438): Support code amalgamation
* [#3381](https://github.com/xmake-io/xmake/issues/3381): Add `xmake test` support
* [#4276](https://github.com/xmake-io/xmake/issues/4276): Support custom scope api
* [#4286](https://github.com/xmake-io/xmake/pull/4286): Add Apple XROS support

### Changes

* [#4284](https://github.com/xmake-io/xmake/issues/4284): Improve builtin includes

### Bugs fixed

* [#4256](https://github.com/xmake-io/xmake/issues/4256): Fix intellisense for vsxmake/c++modules

## v2.8.3

### New features

* [#4122](https://github.com/xmake-io/xmake/issues/4122): Support Lua Debugger (EmmyLua)
* [#4132](https://github.com/xmake-io/xmake/pull/4132): Support cppfront
* [#4147](https://github.com/xmake-io/xmake/issues/4147): Add hlsl2spv rule
* [#4226](https://github.com/xmake-io/xmake/issues/4226): Support sanitizers for package and policy
* Add lib.lua.package module
* Add `run.autobuild` policy
* Add global policies `xmake g --policies=`

### Changes

* [#4119](https://github.com/xmake-io/xmake/issues/4119): Improve to support emcc toolchain and emscripten package
* [#4154](https://github.com/xmake-io/xmake/issues/4154): Add `xmake -r --shallow target` to rebuild target without deps
* Add global ccache storage directory
* [#4137](https://github.com/xmake-io/xmake/issues/4137): Support Qt6 for Wasm
* [#4173](https://github.com/xmake-io/xmake/issues/4173): Add recheck argument to on_config
* [#4200](https://github.com/xmake-io/xmake/pull/4200): Improve remote build to support debugging xmake source code.
* [#4209](https://github.com/xmake-io/xmake/issues/4209): Add extra and pedantic warnings

### Bugs fixed

* [#4110](https://github.com/xmake-io/xmake/issues/4110): Fix extrafiles
* [#4115](https://github.com/xmake-io/xmake/issues/4115): Fix compile_commands generator for clangd
* [#4199](https://github.com/xmake-io/xmake/pull/4199): Fix compile_commands generator for c++ modules
* Fix os.mv fail on window
* [#4214](https://github.com/xmake-io/xmake/issues/4214): Fix rust workspace build error

## v2.8.2

### New features

* [#4002](https://github.com/xmake-io/xmake/issues/4002): Add soname and version support
* [#1613](https://github.com/xmake-io/xmake/issues/1613): Add avx512 and sse4.2 for add_vectorexts
* [#2471](https://github.com/xmake-io/xmake/issues/2471): Add set_encodings to set source/target encodings
* [#4071](https://github.com/xmake-io/xmake/pull/4071): Support the stm8 assembler on the sdcc toolchain.
* [#4101](https://github.com/xmake-io/xmake/issues/4101): Add force includes for c/c++
* [#2384](https://github.com/xmake-io/xmake/issues/2384): Add extrafiles for vs/vsxmake generator

### Changes

* [#3960](https://github.com/xmake-io/xmake/issues/3960): Improve msys2/crt64 support
* [#4032](https://github.com/xmake-io/xmake/pull/4032): Remove some old deprecated apis
* Improve to upgrade vcproj files in tools.msbuild
* Support add_requires("xmake::xxx") package
* [#4049](https://github.com/xmake-io/xmake/issues/4049): Improve rust to support cross-compilation
* Improve clang modules support

### Bugs fixed

* Fix exit all child processes on macOS/Linux

## v2.8.1

### New features

* [#3821](https://github.com/xmake-io/xmake/pull/3821): Add longpath option for windows installer
* [#3828](https://github.com/xmake-io/xmake/pull/3828): Add support for zypper package manager
* [#3871](https://github.com/xmake-io/xmake/issues/3871): Improve tools.msbuild to support for upgrading vsproj
* [#3148](https://github.com/xmake-io/xmake/issues/3148): Support grpc for protobuf
* [#3889](https://github.com/xmake-io/xmake/issues/3889): Support to add library path for add_links
* [#3912](https://github.com/orgs/xmake-io/issues/3912): Add set_pmxxheader to support objc precompiled header
* add_links support library file path

### Changes

* [#3752](https://github.com/xmake-io/xmake/issues/3752): Improve os.getenvs for windows
* [#3371](https://github.com/xmake-io/xmake/issues/3371): Improve tools.cmake to support ninja generator for wasm
* [#3777](https://github.com/xmake-io/xmake/issues/3777): Improve to find package from pkg-config
* [#3815](https://github.com/xmake-io/xmake/pull/3815): Improve tools.xmake to pass toolchains for windows
* [#3857](https://github.com/xmake-io/xmake/issues/3857): Improve to generate compile_commands.json
* [#3892](https://github.com/xmake-io/xmake/issues/3892): Improve to search packages from description
* [#3916](https://github.com/xmake-io/xmake/issues/3916): Improve to build swift program, support for multiple modules
* Update lua runtime to 5.4.6

### Bugs fixed

* [#3755](https://github.com/xmake-io/xmake/pull/3755): Fix find_tool from xmake/packages
* [#3787](https://github.com/xmake-io/xmake/issues/3787): Fix packages from conan 2.x
* [#3839](https://github.com/orgs/xmake-io/discussions/3839): Fix vs_runtime for conan 2.x

## v2.7.9

### New features

* [#3613](https://github.com/xmake-io/xmake/issues/3613): Add `wasm.preloadfiles` configuration for wasm
* [#3703](https://github.com/xmake-io/xmake/pull/3703): Support for conan >=2.0.5

### Changes

* [#3669](https://github.com/xmake-io/xmake/issues/3669): Improve cmake generator to support add_cxflags with the given tool
* [#3679](https://github.com/xmake-io/xmake/issues/3679): Improve `xrepo clean`
* [#3662](https://github.com/xmake-io/xmake/issues/3662): Improve cmake/make generator for lex/yacc project
* [#3697](https://github.com/xmake-io/xmake/issues/3662): Improve trybuild/cmake
* [#3730](https://github.com/xmake-io/xmake/issues/3730): Improve c++modules package installation

### Bugs fixed

* [#3596](https://github.com/xmake-io/xmake/issues/3596): Fix check_cxxfuncs and check_cxxsnippets
* [#3603](https://github.com/xmake-io/xmake/issues/3603): Fix `xmake update`

xmake/modules/private/action/trybuild/xrepo.lua
--!A cross-platform build utility based on Lua
--
-- Licensed under the Apache License, Version 2.0 (the "License");
-- you may not use this file except in compliance with the License.
-- You may obtain a copy of the License at
--
--     http://www.apache.org/licenses/LICENSE-2.0
--
-- Unless required by applicable law or agreed to in writing, software
-- distributed under the License is distributed on an "AS IS" BASIS,
-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-- See the License for the specific language governing permissions and
-- limitations under the License.
--
-- Copyright (C) 2015-present, TBOOX Open Source Group.
--
-- @author      ruki
-- @file        xrepo.lua
--

-- imports
import("core.base.option")
import("core.base.semver")
import("core.project.config")
import("lib.detect.find_file")
import("lib.detect.find_tool")
import("private.action.require.impl.search_packages")

-- get build directory
function _get_buildir()
    return config.buildir() or "build"
end

-- get artifacts directory
function _get_artifacts_dir()
    return path.absolute(path.join(_get_buildir(), "artifacts"))
end

-- detect build-system and configuration file
function detect()

    -- we need xrepo
    local xrepo = find_tool("xrepo")
    if not xrepo then
        return
    end

    -- get package name and version
    local dirname = path.filename(os.curdir())
    local packagename = dirname
    local version = semver.match(dirname)
    if version then
        local pos = dirname:find("v" .. version:rawstr(), 1, true) or dirname:find(version:rawstr(), 1, true)
        if pos then
            packagename = dirname:sub(1, pos - 1)
            if packagename:endswith("-") or packagename:endswith("_") then
                packagename = packagename:sub(1, #packagename - 1):lower()
            end
        end
    end
    packagename = packagename:trim()
    if #packagename == 0 then
        return
    end

    -- search packages
    local result
    local packages_found = search_packages(packagename, {require_version = version and version:rawstr() or nil})
    for name, packages in pairs(packages_found) do
        for _, package in ipairs(packages) do
            if package.name == packagename or packagename:levenshtein(package.name) < 3 then
                result = package
                break
            end
        end
    end
    if result then
        _g.package = result
        if result.version then
            return ("%s %s in %s"):format(result.name, result.version, result.reponame)
        else
            return ("%s in %s"):format(result.name, result.reponame)
        end
    end
end

-- get common configs
function _get_common_configs(argv)
    table.insert(argv, "-y")
    table.insert(argv, "--shallow")
    table.insert(argv, "-v")
    if option.get("diagnosis") then
        table.insert(argv, "-D")
    end
    if config.get("plat") then
        table.insert(argv, "-p")
        table.insert(argv, config.get("plat"))
    end
    if config.get("arch") then
        table.insert(argv, "-a")
        table.insert(argv, config.get("arch"))
    end
    if config.get("mode") then
        table.insert(argv, "-m")
        table.insert(argv, config.get("mode"))
    end
    if config.get("kind") then
        table.insert(argv, "-k")
        table.insert(argv, config.get("kind"))
    end
    if config.get("toolchain") then
        table.insert(argv, "--toolchain=" .. config.get("toolchain"))
    end
    if config.get("vs_runtime") then
        table.insert(argv, "-f")
        table.insert(argv, "vs_runtime='" .. config.get("vs_runtime") .. "'")
    end
end

-- get install configs
function _get_install_configs(argv)

    -- pass jobs
    if option.get("jobs") then
        table.insert(argv, "-j")
        table.insert(argv, option.get("jobs"))
    end

    -- cross compilation
    if config.get("sdk") then
        table.insert(argv, "--sdk=" .. config.get("sdk"))
    end

    -- android
    if config.get("ndk") then
        table.insert(argv, "--ndk=" .. config.get("ndk"))
    end
    if option.get("ndk_sdkver") then
        table.insert(argv, "--ndk_sdkver=" .. option.get("ndk_sdkver"))
    end
    if option.get("android_sdk") then
        table.insert(argv, "--android_sdk=" .. option.get("android_sdk"))
    end
    if option.get("build_toolver") then
        table.insert(argv, "--build_toolver=" .. option.get("build_toolver"))
    end
    if option.get("ndk_stdcxx") then
        table.insert(argv, "--ndk_stdcxx=" .. option.get("ndk_stdcxx"))
    end
    if option.get("ndk_cxxstl") then
        table.insert(argv, "--ndk_cxxstl=" .. option.get("ndk_cxxstl"))
    end

    -- mingw
    if config.get("mingw") then
        table.insert(argv, "--mingw=" .. config.get("mingw"))
    end

    -- msvc
    if config.get("vs") then
        table.insert(argv, "--vs=" .. config.get("vs"))
    end
    if config.get("vs_toolset") then
        table.insert(argv, "--vs_toolset=" .. config.get("vs_toolset"))
    end
    if config.get("vs_sdkver") then
        table.insert(argv, "--vs_sdkver=" .. config.get("vs_sdkver"))
    end

    -- xcode
    if config.get("xcode") then
        table.insert(argv, "--xcode=" .. config.get("xcode"))
    end
    if config.get("xcode_sdkver") then
        table.insert(argv, "--xcode_sdkver=" .. config.get("xcode_sdkver"))
    end
    if config.get("target_minver") then
        table.insert(argv, "--target_minver=" .. config.get("target_minver"))
    end
    if config.get("appledev") then
        table.insert(argv, "--appledev=" .. config.get("appledev"))
    end
end

-- do clean
function clean()
end

-- do build
function build()

    -- get xrepo
    local xrepo = assert(find_tool("xrepo"), "xrepo not found!")

    -- get package info
    local package = assert(_g.package, "package not found!")

    -- do install for building
    local argv = {"install"}
    _get_common_configs(argv)
    _get_install_configs(argv)
    table.insert(argv, "-d")
    table.insert(argv, ".")
    if package.version then
        table.insert(argv, package.name .. " " .. package.version)
    else
        table.insert(argv, package.name)
    end
    os.vexecv(xrepo.program, argv)

    -- do export
    local artifacts_dir = _get_artifacts_dir()
    local argv = {"export"}
    _get_common_configs(argv)
    table.insert(argv, "-o")
    table.insert(argv, artifacts_dir)
    if package.version then
        table.insert(argv, package.name .. " " .. package.version)
    else
        table.insert(argv, package.name)
    end
    os.vexecv(xrepo.program, argv)
    cprint("output to ${bright}%s", artifacts_dir)
    cprint("${color.success}build ok!")
end
